Refactor batch indexing to be resumeable

danielbachhuber opened the pull request:
* [x] Introduce new `SolrPower_Batch_Index` class for handling batch indexing.
* [x] Capture and resume index based off `paged`
* [x] Refactor `wp solr index` to use batch index class and provide more verbosity.
* [x] Refactor admin UX to use batch index class.
* [x] Remove old `wp solr index` command.
* [x] Remove old code associated with admin indexing UX.
* [ ] Add tests around the batch index process.
* [x] Fire the `solr_power_index_all_finished` action at the end of the indexing process.

Fixes #191

danielbachhuber commented:
> However, the admin UI hung with 81 posts left.

Ah, probably because I still need to:

![image](https://cloud.githubusercontent.com/assets/36432/23270232/0b30bc3c-f9a8-11e6-9030-9043c5379a39.png)

ataylorme commented:
Screenshots

![screen shot 2017-02-23 at 11 03 11 am](https://cloud.githubusercontent.com/assets/2133004/23270242/d44e8dfa-f9b8-11e6-8e94-80efc2f1adc6.png)
![screen shot 2017-02-23 at 11 10 20 am](https://cloud.githubusercontent.com/assets/2133004/23270243/d452ef1c-f9b8-11e6-9a64-0ee28e0b2893.png)

ataylorme commented:
Hah! I totally missed that. Once that's done I'll test again and approve the PR.

danielbachhuber commented on includes/class-solrpower-batch-index.php:
```diff
@@ -0,0 +1,265 @@
+	public function increment_page() {
+		$this->query_args['paged'] = $this->query_args['paged'] + 1;
+		delete_option( $this->batch_cache_key );
+		add_option( $this->batch_cache_key, $this->query_args['paged'], null, 'off' );
```
> > Are you using `delete_option()` and `add_option()` rather than `update_option()` to disable autoloading in cases where the option doesn't exist and needs to be added?

Yes

> If so the `$autoload` parameter should be `false` or `'no'`, not `'off'`

D'oh! Fixed bd036ac

danielbachhuber commented on includes/class-solrpower-batch-index.php:
```diff
@@ -0,0 +1,265 @@
+	 * @return bool
+	 */
+	public function have_posts() {
+		if ( ! empty( $this->post_ids ) ) {
```
Sure, 911064a

danielbachhuber commented on includes/class-solrpower-batch-index.php:
```diff
@@ -0,0 +1,265 @@
+	 * Increment value for 'paged', and update its value in the cache
+	 */
+	public function increment_page() {
+		$this->query_args['paged'] = $this->query_args['paged'] + 1;
```
Sounds good, 60252a2

danielbachhuber commented:
> Hah! I totally missed that. Once that's done I'll test again and approve the PR.

I've updated the UI for indexing, and added a completion message:

![indexing-ux](https://cloud.githubusercontent.com/assets/36432/23286096/3390ddb4-f9ea-11e6-8081-4f600d43c4f0.gif)

I still want to add tests for the batch index class, but can do so in a follow-up PR #217

This PR is good to 🚢

ataylorme commented:
@danielbachhuber this is looking good to me. @joshkoenig any objection to merging?
